Make zk2 runnable

Risca opened the pull request:
Updated enough to build znake, zaf, and run zk2 (with monitors disabled).

It is necessary to search/replace or spoof example.com to get it to work. It is designed for a local pip mirror. I found it next to impossible to make it work in both an offline environment and one where it's connected to the internet. This potentially should work, if you just redirect example.com to a pypi mirror.

To make the docker images work, you have to manually install them locally and then run znake with `--no-pull` (and possibly `--network=host`)

andni233 commented on znake/activate:
```diff
@@ -1,9 +1,18 @@
 # Source this file to create and activate a Znake virtual environment
-python3.6 -m venv .venv
+python3.8 -m venv .venv
```
I think I made a working verision of Znake and friends available here ages ago:
https://github.com/ZenDevelopmentEcosystem

I'm available if you want to make an attempt at resurrecting this stuff.

Risca replied:
I was thinking of forking this repo a bunch of times and then move files around in each repo, to maintain the history and origin ^^

Risca replied:
I have an ambition of resurrecting at least some of these projects :D Looking at the code, I can see that a lot of love and effort has been put into this code, and, frankly, I haven't found anything remotely as advanced and yet simple to use, anywhere.

per-bohlin replied:
All new development in these products are done at https://github.com/ZenDevelopmentEcosystem.
I hope to have some more time to work soon this, especially k2 and make it use-able in non-Zenterio context.
Any help is greatly appriciated. Feel free to reach out and discuss future roadmap.

Risca replied:
What would be the best forum for discussion? Feels a bit clunky in a review.
Maybe create an issue for each project to be imported? I noticed that neither zaf nor k2-core have any actual code in them

Risca commented:
Will make separate pull requests at https://github.com/ZenDevelopmentEcosystem

Risca commented:
Ok, turns out that no changes were needed to https://github.com/ZenDevelopmentEcosystem/znake, and the other repos have no code.

What to do now?

per-bohlin commented:
Sorry for my late reply.

As usual things has come up :(, so I will not be able to work on k2-core and zaf this weekend as I had hopped.

So I recommend that until further notice:
* All future modifications to znake should be done in https://github.com/ZenDevelopmentEcosystem/znake
* Modifications to zaf and k2 can be done to https://github.com/Zenterio/opensourcelib. I will make sure your changes are taken into account once we release first official versions in ZenDevelopmentEcosystem.

Risca commented:
Alright, then I'll reopen this pull request :wink: I've also fixed a few more things. With these changes, all unit tests for k2 and `znake check` for _zaf_ passes :smiley_cat:

Risca commented on zaf/zaf/component/factory.py:
```diff
@@ -144,6 +152,10 @@ def call(self, callable, scope, *args, fixated_entities=None, extra_req=None, **
 
+        for name in pre_instantiated or {}:
+            instance = pre_instantiated[name]
+            scope.register_instance(_InstanceId(instance.__class__, args, kwargs), instance)
```
I'm not 100% sure about this line, but it works :shrug:

per-bohlin replied:
What problem will this solve?

Risca replied:
Hmm... For some reason I can't reproduce the issue now :confused: I kept hitting this RunnerComponentException in _k2/k2/runner/components.py_:
```python
    def __init__(self, test_case_definition=None):
        if test_case_definition is None:
            raise RunnerComponentException(
                'The TestContext component can only be pre-instantiated by the test runner and can only be used '
                'on the test scope.')
```

Risca replied:
This is how the call to the component factory looks like:
```python
        def run_test_case(*args, **kwargs):
...
                            self.component_factory.call(
                                current_test_case.run,
                                scope,
                                *args,
                                extra_req=extra_req,
                                pre_instantiated=self.create_pre_instantiated_components(
                                    current_test_case),
                                **kwargs)
...
    def create_pre_instantiated_components(self, test_case_definition):
        """Create the pre-instantiated TestContext instance."""
        return {'TestContext': TestContext(test_case_definition)}
```

Risca replied:
Commit removed in latest diff. Issue is fixed by upstream changes =)

Risca commented on k2/requirements-dev.txt:
```diff
@@ -10,7 +10,6 @@ flake8-string-format==0.2.3
 flake8==3.7.6
 flake8_tuple==0.2.13
 isort==4.3.0
-pdoc==0.3.2
```
I ran into pip problems with package and the solution seems to be to switch to pdoc3, but I also found no references to pdoc so I removed it instead.

Risca commented on k2/requirements.txt:
```diff
@@ -13,26 +12,27 @@ addons/k2-orchestration
 wakeonlan==1.1.6
 websocket-client==0.55.0
-zenterio-zaf==0.28.1
+zenterio-zaf
```
I had some problems with `znake pypi` producing _zenterio-zaf-1.0.0.post0_ packages which doesn't match "1.0.0" requirement. I think this is fine for now at least.

Risca commented on k2/addons/k2-metrics/metrics/metrics.py:
```diff
@@ -159,16 +159,19 @@ def __getitem__(self, name):
+                    if isinstance(metric, MetricsNamespace):
+                        yield metric
+        except StopIteration:
+            return
```
This is actually a fix for python 3.8, where `StopIteration` started causing runtime exception.

Risca commented on k2/k2/reports/junit/test/test_writer.py:
```diff
@@ -4,6 +4,8 @@
 import unittest
 from textwrap import dedent
 
+from xmldiff import main
```
Different python versions would output xml documents with attributes listed in different order, so I opted for using a 3rd party library to compare xml documents.

Risca commented on k2/k2/reports/testng/test/test_writer.py:
```diff
@@ -4,6 +4,8 @@
 import unittest
 from textwrap import dedent
 
+from xmldiff import main
```
Same issue as for junit unittests

Risca commented on zaf/zaf/application/metadata.py:
```diff
@@ -108,11 +108,7 @@ def _init_extension_classes(self):
-                            for endpoint in sorted(
-                                ext_class.endpoints_and_messages.keys(), key=lambda e: e.name)
-                        ],
+                        sorted(ext_class.endpoints_and_messages.keys(), key=lambda e: e.name),
```
flake8 on Ubuntu server 18.04 complained about unnecessary list comprehension

Risca commented on zaf/zaf/builtin/click/click.py:
```diff
@@ -215,10 +215,7 @@ def __init__(
-            command for command in get_applicable_commands(all_commands, self._application_context)
-        ]
-
+        self._all_commands = list(get_applicable_commands(all_commands, self._application_context))
```
flake8 on Ubuntu server 18.04 complained about unnecessary list comprehension

Risca commented on zaf/zaf/extensions/manager.py:
```diff
@@ -250,7 +250,7 @@ def is_command_capability(extend):
 
             if (command is None or command in commandid_extends
-                    or command in {cmd.name for cmd in commandid_extends}
+                    or command in (cmd.name for cmd in commandid_extends)
```
flake8 on Ubuntu server 18.04 complained about unnecessary set comprehension

Risca commented:
Btw, 'make systest' in k2 fails with complaints about '--suts-id'. I haven't had time to debug that yet but I don't think it's blocking this review

per-bohlin commented on k2/k2/reports/junit/test/test_writer.py:
```diff
@@ -138,7 +153,6 @@ def test_run_verdict_mapped_to_test_case(self):
 def no_exception(time_index, verdict, name_suffix=''):
     return dedent(
         f"""\
-        <?xml version="1.0" encoding="utf-8"?>
```
what caused the removal of the XML prolog?

Risca replied:
The `xmldiff` library would give me an exception if I left it there. I believe I can restore it and sprinkle the code with `encode('utf-8')` instead. It seems those are the 2 options that xmldiff give me:
* give it `bytes()` with encoding tag, or
* give it `str()` without encoding tag.

I didn't know which one was best, so I chose one of them.

Side note: it seems the _junit_ report writer produce a `str()`, while the _testng_ report writer produce a `bytes()`. Again, I don't know which one is most correct. Feels like it's out of scope for this pull request at least.

Risca replied:
Prolog restored in latest diff, and strings encoded to python `bytes()` objects instead

per-bohlin commented:
The divergence in interfaces made me auspicious and I started to dig through the ZAF and K2 code.
I found newer version of ZAF that had not been committed here, that align with the version of K2 committed in this repository. It fixes some of the issues you have encountered and more.
Hence we have competing changes.
I believe the changes I have are more complete to the areas they touch and could benifit ZAF in the long run.
How do you feel about letting me update ZAF first and you rebase on top of it?
You are more familiar with your changes than I am with the changes I have access to (they where written by other members of my former team), and I'm afraid I will make a mess of it if we do it the other way around.
I will continue digging to see if I can find additional updates to K2 as well.

Risca commented:
I have no problem re-basing! I have more trust in you and your former team members than I have in my fixes :laughing:

Risca commented:
Branch rebase and updated. Please review and accept if you find it acceptable =)

Risca commented on k2/addons/k2-sutevents/sutevents/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from zaf.messages.message import MessageId
+
+LOG_LINE_RECEIVED = MessageId('LOG_LINE_RECEIVED', """Log line received""")
```
I can't remember exactly what sutevents did, but at least this `MessageId()` is needed to use the zserial component.

Risca replied:
If you happen to find k2-sutevents somewhere upstream, I'm happy to rebase again :smiley:

per-bohlin replied:
I have a couple of more addons, including k2-sutevents.
I didn't think they where dependent, so I thought I would have a couple of days to fix them up. I'll commit them shortly.

per-bohlin replied:
I've just merged #10 that contains k2-sutevents.
I'm sorry for the hassle of having to rebase again.

Risca commented:
It seems that `znake static` in _zaf_ will give me wildly different results on Ubuntu 18.04 (python 3.6) vs. Arch Linux (python 3.9), so I'm going with the path of least resistance and only fixing stuff that came up in Arch Linux. It resulted in the least amount of changes, whereas `znake format` on Ubuntu 18.04 resulted in 38 changed files :S

Risca commented:
* Systest and unittest passed for both znake and zaf, tested on Ubuntu server 18.04 (python 3.6) and Arch Linux (python 3.9)
* Unittest passed for k2 on Ubuntu server 18.04 (python 3.6) and Arch Linux (python 3.9)

Risca commented:
With the latest commits, even k2 systests are passing :smiley:

Risca commented:
Fixed some problems reported by `make check`. Now it runs without any errors :smiley_cat: Tested on both Arch Linux and Ubuntu Server 18.04 :wink:

per-bohlin commented:
Given the large number of changes and comments in this merge-request. Consider if it would be cleaner to create a new pull request for the latest rebase. If you find additional code completely missing, let me know I might be able to dig it up.

Risca commented:
I'm kinda missing the `gudepowermeter` and `gudepowerswitch` from the k2-power addon. I have disabled some of the systests there because these 2 are missing. If I remember correctly, those 2 components just implement the public API from here: http://wiki.gude.info/EPC_HTTP_Interface

Risca commented:
Also, the multirunner part of the `k2-consolerunner` addon is missing. Because of this, the systests for consolerunner are disabled.

Risca commented:
I've rebased and split up this pull request into a few smaller pull requests. Hopefully, this will make it easier to work with. Closing this one in favor of #11 and #13

Risca commented:
> Given the large number of changes and comments in this merge-request. Consider if it would be cleaner to create a new pull request for the latest rebase. If you find additional code completely missing, let me know I might be able to dig it up.

I seem to remember an addon that allowed you to transfer files to the SUT. Maybe you can see if that one is available?
